### ISSUE TITLE
[New Concept Docs]: Higher Order Functions

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#3103]
This issue describes how to implement the `higher-order functions` in Python  **concept docs**.
You can find the related **concept exercise** issue [here](https://github.com/exercism/python/issues/3104)

If you have not yet contributed to concept documents, this issue will require some upfront reading to give you the needed background knowledge.

## ✅   Getting started

**`Please please read the docs before starting.`** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. 

**General Contributing Docs:**

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | - [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

**Documents on Language Tracks and Concepts:**

- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 

<br>

## 🎯   Goal

These concept docs are meant to teach a deeper understanding/use/creation of  `higher-order functions`  in Python. 

<br>

## 💡   Learning objectives

- Understand and create `higher-order functions`  in Python
- Create `functions` used as arguments to other `functions`
- Create `functions` that return  `functions` 
- Learn about `currying`
- Understand and create  `closures`  -- where an inner function has access to a variable from it's enclosing scope (_the outer function_) that has completed its execution.

<br>

## 🚫  Out of scope

<details>
<summary><b>Concepts & Subjects that are Out of Scope</b> (<em>click to open</em>)</summary>
<br>

This exercise assumes an understanding of its prerequisites, so it is not necessary to go into scoping, function special attributes,  function arguments, or other concepts "higher" in the syllabus "tree".

- `comprehensions`
- `decorators` (_these will have their own exercise_)
- `functools` (_this will get its own exercise_)
- `functools.wraps`
- `generators`
- `lambda`, `anonymous functions`
- `map()`, `filter()`, and `reduce()` (_these will get their own exercise_)
-  `nonlocal` & `locals()`
-  `global` & `globals()`
-   `argument unpacking`
-  `recursion`

</details>

<br>

## 🤔  Concepts

<details>
<summary><b>Concepts and Related Concepts this Concept Covers</b> (<em>click to open</em>)</summary>
<br>

- `closures`
- `currying`
- `functions`, `higher-order functions`
- `functions as arguments`
- `functions as returns`

</details>

<br>

## ↩️   Prerequisites

These are the concepts/concept exercises the student should be familiar with before taking on/learning this concept.  

<details>
<summary>Prereqs  (<em>click to open</em>)</summary>
<br>

- `basics`
- `booleans`
- `comparisons`
- `dicts`
- `dict-methods`
- `functions`
- `function-arguments`
-  `argument unpacking`
- `iteration`
- `lists`
- `list-methods`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `string-methods`
- `tuples`

</details>

<br>  

## 📚  Resources to refer to

<details>
<summary><b>Resources</b>  (<em>click to open</em>)</summary>
<br>

- [Python Docs: Python Scopes and Namespaces](https://docs.python.org/3/tutorial/classes.html#python-scopes-and-namespaces)
- [Python Docs: Defining Functions](https://docs.python.org/3/tutorial/controlflow.html#defining-functions)
- [Dan Bader:  Python s Functions are First-Class](https://dbader.org/blog/python-first-class-functions)
- [Functions as Objects in Python](https://medium.com/python-pandemonium/function-as-objects-in-python-d5215e6d1b0d)
- [Composing Programs: Higher-Order Functions](https://composingprograms.com/pages/16-higher-order-functions.html)
- [Tobias Kohn: Closures in Python](https://tobiaskohn.ch/index.php/2018/08/06/closures-in-python/)
- [zetcode: Python Closures](https://zetcode.com/python/python-closures/#:~:text=Three%20characteristics%20of%20a%20Python,returned%20from%20the%20enclosing%20function)
- [Sagnick:  Python is the Haskell you Never Knew You Had: Currying](https://sagnibak.github.io/blog/python-is-haskell-currying/)
- [Oreilly:  Associating Parameters with a Function (Currying)](https://www.oreilly.com/library/view/python-cookbook/0596001673/ch15s08.html)
- [Real Python:  Thinking Recursively in Python](https://realpython.com/python-thinking-recursively/)
- [Train Your Brain to Think Recursively](https://towardsdatascience.com/train-your-mind-to-think-recursively-in-5-steps-8f85c0e0eb81)

</details>
<br>

##  📁  Files to Be Created

<details open>
<summary><b>File Detail for these Concept Docs</b></summary>
<table>
<td>
<br>

Please see the following for more details on these files:  [concepts](https://exercism.org/docs/building/tracks/concepts)

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)

  -  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  -  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

* ### Concept `about.md`

  For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-aboutmd)

  - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-aboutmd)

  * This can also be a summary/paraphrase of the  `about.md` document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the associated concept or practice exercises.   It should contain a good summation of the concept, but not go into lots of detail.

* ### Concept `.meta/config.json` Entries
 
  For more information, see [Concept `.meta/config.json`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-metaconfigjson)

  * This file is likely already stubbed out.  Remember to add a concept **blurb** of less than 350 characters.  Please also add your GitHub username to the "authors" array, and any contributor GitHub usernames to the "contributors" array.

<br>
</td>
</table>
</details>

## 🎶  Implementation Notes

- Example code  should **only use syntax & concepts introduced within these docs or one of the prerequisite concept exercises or documents.** Where possible, please use REPL formatting, unless you are demonstrating pseudo code or a long code block.
  Please **do not use** syntax not previously covered in prerequisite topics or exercises.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

<br>

##  🆘  Next Steps & Getting Help

1.  **`If you'd like to work on this issue, comment saying "I'd like to work on this"`** (_there is no real need to wait for a response, just go ahead, we'll assign you and put a `[claimed]` label on the issue_).
2. If you have any questions while implementing, please post the questions as comments in here, or contact one of the maintainers on our Slack channel.
